sys/newlib_syscalls_default/syscalls.c

```diff
@@ -142,7 +142,7 @@ static const struct heap heaps[NUM_HEAPS] = {
 /**
  * @brief Initialize NewLib, called by __libc_init_array() from the startup script
  */
-void _init(void)
+__attribute__ ((weak)) void _init(void)
```

Comment:
Yes, there are indeed 2 options.

But first, I should give more details about the original problem. ;)
```
/home/gdo/Applications/toolchain/arm-gnu-toolchain-13.2.Rel1-x86_64-arm-none-eabi/bin/../lib/gcc/arm-none-eabi/13.2.1/../../../../arm-none-eabi/bin/ld: /home/gdo/Applications/toolchain/arm-gnu-toolchain-13.2.Rel1-x86_64-arm-none-eabi/bin/../lib/gcc/arm-none-eabi/13.2.1/../../../../arm-none-eabi/lib/thumb/v7e-m+fp/hard/libstdc++_nano.a(globals_io.o): in function `_GLOBAL__sub_I.00090__ZSt3cin':
globals_io.cc:(.text.startup._GLOBAL__sub_I.00090__ZSt3cin+0x1c): undefined reference to `__dso_handle'
```
The `__dso_handle` should be used by `int __cxa_atexit(void (*func) (void *), void * arg, void * dso_handle);` when exiting or unloding shared object.
Such case is not relevant for RIOT, thus lets disable this with `-fno-use-cxa-atexit`. ==>> Same error `undefined reference to `__dso_handle'`.

So there, I got 2 options.

Option 1:
Just add weak `__dso_handle` definition:
```diff
diff --git a/sys/newlib_syscalls_default/syscalls.c b/sys/newlib_syscalls_default/syscalls.c
index d06db6da94..68639a6af1 100644
--- a/sys/newlib_syscalls_default/syscalls.c
+++ b/sys/newlib_syscalls_default/syscalls.c
@@ -144,10 +144,12 @@ static const struct heap heaps[NUM_HEAPS] = {
 #endif
 };
 
+void *__dso_handle __attribute__((weak)) = NULL;
+
 /**
  * @brief Initialize NewLib, called by __libc_init_array() from the startup script
  */
```

Option 2 (done by this commit):
Let the stdlib, here the newlib, handling start files by removing `-nostartfiles` flag from `LINKFLAGS` variable in my application Makefile.
```Makefile
LINKFLAGS:=$(filter-out -nostartfiles,$(LINKFLAGS))
```
But there, as newlib is also defining `void _init(void)` and `void _fini(void)`, I had to define them as `weak` in RIOT.

I choose the option 2 just because IMHO, RIOT should force the user to use its own start files functions if the user decided to use newlib ones.

However, I am clearly not an expert of all of this, so if option 1 seems better, or if there is a third option, I'm all ok to fix it. :wink:

Reply:
> Just add weak `__dso_handle` definition:

This is what you get when you do

    USEMODULE += cpp

via the  `cpp_new_delete` module

Reply:
Yes but only if I use `cpp11-compat` instead of libstdcpp. And we do not use cpp11-compat.

`FEATURES_REQUIRED += cpp` does not seem to interfer here, do you confirm ? :thinking:

Reply:
If I set `USEMODULE += cpp_new_delete`, it builds fine.

Reply:
huh, I thought #20348 would have fixed that

Reply:
I don't know the implications of using newlib start files either, so let's better leave it as it is then.

Reply:
top, thx @benpicco @chrysn. Done for this one.